### PR TITLE
+ ref and elision

### LINKED ISSUE #61: scope/borrow 2
1. [ref.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/scope/borrow/ref.md)
2. [elision.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/scope/lifetime/elision.md)

### PR DESCRIPTION
**Pull request by HieuTranV**
Closes: #61

### CODE REVIEW

**Comment by HieuTranV** [comment]
closes: #61

File: src/scope/borrow/ref.md [hunk 1]
```diff
@@ -1 +1,55 @@
 # The ref pattern
+
+Khi thực hiện lệnh gán hoặc trích xuất giá trị (của một struct) thông qua lệnh  `let`, từ khóa `ref` có thể được sử dụng lấy ra thể hiện của các trường trong một biến kiểu struct/tuple. Ví dụ dưới đây là một vài tình huống có thể sử dụng kỹ thuật:
```
Comment (HongThaiPham) [inline src/scope/borrow/ref.md]:
```suggestion
Khi thực hiện matching hoặc destructuring thông qua `let`, từ khóa `ref` có thể được sử dụng lấy tham chiếu của các trường trong một biến kiểu struct/tuple. Ví dụ dưới đây là một vài tình huống có thể hữu ích:
```

File: src/scope/borrow/ref.md [hunk 2]
```diff
@@ -1 +1,55 @@
...
+    let ref ref_c1 = c;
+    let ref_c2 = &c;
+
+    println!("ref_c1 equals ref_c2: {}", *ref_c1 == *ref_c2);
+
+    let point = Point { x: 0, y: 0 };
+
+    // `ref` cũng có thể được dùng trong trường hợp cần trích xuất giá trị từ biến kiểu struct 
+    let _copy_of_x = {
+        // `ref_to_x` là một thể hiện của trường `x` bên trong biến `point`.
```
Comment (HongThaiPham) [inline src/scope/borrow/ref.md]:
```suggestion
        // `ref_to_x` là một tham chiếu đến trường `x` bên trong biến `point`.
```

File: src/scope/borrow/ref.md [hunk 3]
```diff
@@ -1 +1,55 @@
...
+        let Point { x: ref ref_to_x, y: _ } = point;
+
+        // Trả về giá trị của trường `x` trong biến `point`.
```
Comment (HongThaiPham) [inline src/scope/borrow/ref.md]:
```suggestion
        // Trả về bản sao của trường `x` trong biến `point`.
```

File: src/scope/borrow/ref.md [hunk 4]
```diff
@@ -1 +1,55 @@
...
+        *ref_to_x
+    };
+
+    // Một bản sao có thể thay đổi giá trị của biến `point`
+    let mut mutable_point = point;
+
+    {
+        // Từ khóa `ref` có thể đi cùng từ khóa `mut` để tạo ra một thể hiện có thể thay dổi giá trị.
```
Comment (HongThaiPham) [inline src/scope/borrow/ref.md]:
```suggestion
        // Từ khóa `ref` có thể đi cùng từ khóa `mut` để tạo ra một tham chiếu có thể thay đổi giá trị.
```

File: src/scope/borrow/ref.md [hunk 5]
```diff
@@ -1 +1,55 @@
...
+        let Point { x: _, y: ref mut mut_ref_to_y } = mutable_point;
+
+        // Thay đổi trường `y` của `mutable_point` thông qua thể hiện có thể thay đổi giá trị (vừa được tạo ở trên).
```
Comment (HongThaiPham) [inline src/scope/borrow/ref.md]:
```suggestion
        // Thay đổi trường `y` của `mutable_point` thông qua tham chiếu có thể thay đổi giá trị (vừa được tạo ở trên).
```

File: src/scope/borrow/ref.md [hunk 6]
```diff
@@ -1 +1,55 @@
...
+        *mut_ref_to_y = 1;
+    }
+
+    println!("point is ({}, {})", point.x, point.y);
+    println!("mutable_point is ({}, {})", mutable_point.x, mutable_point.y);
+
+    // Một biến tuple có thể thay đổi giá trị có giá trị là một Pointer (điểm)
```
Comment (HongThaiPham) [inline src/scope/borrow/ref.md]:
```suggestion
    // Một biến tuple có thể thay đổi giá trị chứa một Pointer (điểm)
```

File: src/scope/borrow/ref.md [hunk 7]
```diff
@@ -1 +1,55 @@
...
+    let mut mutable_tuple = (Box::new(5u32), 3u32);
+    
+    {
+        // Truy xuất vào biến `mutable_tuple` để thay đổi giá trị của `last`.
```
Comment (HongThaiPham) [inline src/scope/borrow/ref.md]:
```suggestion
        // Destructure biến `mutable_tuple` để thay đổi giá trị của `last`.
```

File: src/scope/lifetime/elision.md [hunk 8]
```diff
@@ -1 +1,42 @@
 # Elision
+
+Trường hợp các biến tồn tại suốt vòng đời (của một hàm) là rất phổ biến và vì thế hệ thống (borrow checker) cho phép bạn bỏ qua việc đánh dấu lifetime để tiết kiệm thời gian nhập và để dễ đọc hơn. Cơ chế Elision tồn tại trong Rust bởi vì các trường hợp như thế này quá phổ biến.
```
Comment (HongThaiPham) [inline src/scope/lifetime/elision.md]:
```suggestion
Trường hợp các biến tồn tại suốt vòng đời (của một hàm) là rất phổ biến và vì thế trình kiểm tra mượn (borrow checker) cho phép bạn bỏ qua việc đánh dấu lifetime để tiết kiệm thời gian nhập và để dễ đọc hơn. Điều này được gọi là Elision. Cơ chế Elision tồn tại trong Rust bởi vì các trường hợp như thế này quá phổ biến.
```

File: src/scope/lifetime/elision.md [hunk 9]
```diff
@@ -1 +1,42 @@
...
+
+Đoạn code dưới đây sễ đưa ra một vài ví dụ về cơ chế Elision. Để hiểu rõ hơn về cơ chế này, bạn có thể đọc tại [đây][lifetime ellision]: 
+```rust,editable
+// `elided_input` và `annotated_input` có những đặc trưng giống nhau
```
Comment (HongThaiPham) [inline src/scope/lifetime/elision.md]:
```suggestion
// `elided_input` và `annotated_input` có mô tả (function signatures) giống nhau
```

File: src/scope/lifetime/elision.md [hunk 10]
```diff
@@ -1 +1,42 @@
...
+// vì lifetime của `elided_input` được trình biên dịch xác định:
+fn elided_input(x: &i32) {
+    println!("`elided_input`: {}", x);
+}
+
+fn annotated_input<'a>(x: &'a i32) {
+    println!("`annotated_input`: {}", x);
+}
+
+// Tương tự, `elided_pass` và `annotated_pass` cũng có những đặc trưng giống nhau
```
Comment (HongThaiPham) [inline src/scope/lifetime/elision.md]:
```suggestion
// Tương tự, `elided_pass` và `annotated_pass` cũng có những mô tả giống nhau
```

File: src/scope/lifetime/elision.md [hunk 11]
```diff
@@ -1 +1,42 @@
...
+// vì lifetime được tự động thêm vào `elided_pass`:
```
Comment (HongThaiPham) [inline src/scope/lifetime/elision.md]:
```suggestion
// vì lifetime được ngầm định thêm vào `elided_pass`:
```

**Review by HongThaiPham — APPROVED** [review]

**Pull request merged**